Development/schedule rework

CptHouston opened the pull request:
-- added new entity (scheduled event)
-- schedule entity renamed to EventOccurnce
-- post schedule endpoint updated

//database updated accordingly

CptHouston committed:
Create dotnet-tools.json

CptHouston committed:
-- reworked postSchedule endpoint
-- currently only weekly pattern available

CptHouston committed:
Revert "Create dotnet-tools.json"

This reverts commit 34e5a519b4dfb4daab5f0549a28a34c558e8f05a.

CptHouston committed:
Delete profile.arm.json

Hlib-mats commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventOccuranceStorageParser.cs:
```diff
@@ -0,0 +1,76 @@
+                    break;                
+                case PatternType.RelativeMonthly:
+                case PatternType.Daily:
+                case PatternType.AbsoluteMonthly:
```
add break or //break if u didn't forget them

Hlib-mats commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,187 @@
+
+                if (scheduleDTO == null)
+                {
+                    return Result<EventOccurenceDTO>.GetError(ErrorCode.NotFound, "UpdateScheduleDto is null");
```
maybe ErrorCode should be UnProcessable entity

CptHouston replied:
ErrorCode updated

Hlib-mats commented on CharlieBackend.Core/DTO/Schedule/EventOccurenceDTO.cs:
```diff
@@ -7,23 +7,25 @@
         [Required]
         public long StudentGroupId { get; set; }
 
         [Required]
-        [DataType(DataType.Time)]
-        public TimeSpan LessonStart { get; set; }
+        public DateTime EventStart { get; set; }
+
+        [Required]        
+        public DateTime EventFinish { get; set; }
```
we should add [DataType(DataType.Time)] attribute i think

Hlib-mats commented on CharlieBackend.Core/Entities/EventOccurence.cs:
```diff
@@ -0,0 +1,37 @@
+
+        public virtual StudentGroup StudentGroup { get; set; }
+
+        [Required]
+        public DateTime EventStart { get; set; }
+
+        public DateTime? EventFinish { get; set; }
```
also here should add [DataType(DataType.Time)]

CptHouston committed:
-- added methods for other patterns

CptHouston committed:
Merge branch 'development/scheduleRework' of https://github.com/CptHouston/WhatBackend into development/scheduleRework

CptHouston committed:
Conflict check

CptHouston committed:
Conflict

CptHouston committed:
Merge branch 'dev' into development/scheduleRework

CptHouston committed:
Updated ScheduledEventHandler

CptHouston committed:
Merge branch 'development/scheduleRework' of https://github.com/CptHouston/WhatBackend into development/scheduleRework

CptHouston committed:
UpdatedDB scripts

CptHouston committed:
-- bugFix

CptHouston committed:
-- updated parser to increase readability

mahotskuy commented on CharlieBackend.Api/Controllers/SchedulesController.cs:
```diff
@@ -69,26 +69,38 @@ public async Task<ActionResult<List<ScheduleDto>>> GetAllSchedules()
         }
 
+        /// <summary>
+        /// Кeturns all events that start or end in the given interval
```
Rename Keturns

CptHouston replied:
Updated

mahotskuy commented on CharlieBackend.Api/Controllers/SchedulesController.cs:
```diff
@@ -69,26 +69,38 @@ public async Task<ActionResult<List<ScheduleDto>>> GetAllSchedules()
+        /// </summary>
+        /// <response code="200">Successful return of schedules list</response>
+        /// <response code="HTTP: 404, API: 0">Error, start date is later than finish date</response>
+        public async Task<ActionResult<List<EventOccurenceDTO>>> GetEventsByDateAsunc(DateTime startTime, DateTime finishTime)
```
What uses case for using this endpoint?

CptHouston replied:
For Update and Delete method discussion please check #364

PRs parted to decrease updated code amount. Endpoint refactored to return filtered scheduled events

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,250 @@
+            if (scheduleEntity != null)
+            {
+                var mappedSchedule = _mapper.Map<EventOccurenceDTO>(scheduleEntity);
+                await _unitOfWork.EventOccurenceRepository.DeleteAsync(id);
```
Should we delete created events if we delete occurrences?

CptHouston replied:
For Update and Delete method discussion please check #364

PRs parted to decrease updated code amount. Currently there is an option to delete all events or some events. All events are removed if event occurrence is deleted by delete cascade db constraint

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,250 @@
+                _mapper.Map<IList<EventOccurenceDTO>>(schedulesOfGroup));
+        }
+
+        public async Task<Result<EventOccurenceDTO>> UpdateStudentGroupAsync(long scheduleId, UpdateScheduleDto scheduleDTO)
```
Rename method name.
Does UpdateScheduleDto contains schduleId? We can pass then only UpdateScheduleDto and get Id from there.

CptHouston replied:
For Update and Delete method discussion please check #364

PRs parted to decrease updated code amount. Method name updated

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,250 @@
+            }
+            catch
+            {
+                _unitOfWork.Rollback();
```
I don't think we have to rollback anything here. Updating of entity executed on CommitAsync method.  If there will be an error then we just simply don't save the data.

CptHouston replied:
For Update and Delete method discussion please check #364

PRs parted to decrease updated code amount. Rollback removed during refactoring

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventOccuranceStorageParser.cs:
```diff
@@ -0,0 +1,154 @@
+            return data.result;
+        }
+
+        internal static PatternForCreateScheduleDTO GetFullDataFromStorage(long source)
```
Rename return type to ScheduleCreationPatternDto

CptHouston replied:
Updated

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduledEventsHandler.cs:
```diff
@@ -0,0 +1,143 @@
+                startDate = source.EventStart
+            };
+
+            switch (source.Pattern)
```
Extract switch case to separate method. Each case should return the result of method like GetDailyEvent() 
switch:
case PatternTepe.Daily => getDailyEvent(source)

mahotskuy replied:
Please use the strategy pattern here. Make this class non static and rename to SheduledEventProviderFactory.
 Create interface IShceduledEventProvider with IEnumerable<ScheduledEvent> GetEvents (source).
Then create switch case method that returns appropriate strategy
new DailyScheduledEventProvider(context)
Then in ScheduleService using of this method could be next
_scheduledEventsProvider.Get(source.Type).GetEvents(source)

mahotskuy commented on CharlieBackend.Core/DTO/Schedule/CreateScheduleDTO/OccurenceRange.cs:
```diff
@@ -0,0 +1,15 @@
+        [Required]
+        public DateTime StartDate { get; set; }
+
+        public DateTime? FinishDate { get; set; }
```
I think FinishDate should be not nullable

CptHouston replied:
Updated, required attribute added

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,250 @@
+
+                foundSchedule.Pattern = updatedEntity.Pattern;
+
+                foundSchedule.Storage = updatedEntity.Storage;
```
I think we have to delete AllScheduledEvents and then recreate them.

CptHouston replied:
For Update and Delete method discussion please check https://github.com/ita-social-projects/WhatBackend/pull/364

PRs parted to decrease updated code amount. Update schedule in progress (draft)

CptHouston committed:
Added inharitance
-- updated bug with swagger

CptHouston committed:
CreateScheduleAsync refactoring
added subfolders + removed provate methods

CptHouston committed:
-- bugfixed

CptHouston committed:
-- updated naming

TetianaDomanska-dev commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerAbsoluteMonthly.cs:
```diff
@@ -0,0 +1,34 @@
+
+        public override void UpdateTime(ref DateTime starttime, ref DateTime finishDate)
+        {
+            starttime = starttime.AddDays(_pattern.Interval);
```
```suggestion
            startDate = startDate.AddDays(_pattern.Interval);
```
Maybe it will be more logic with such changes.

CptHouston replied:
Updated

CptHouston committed:
-- GetFullDataFromStorage set to public access

CptHouston committed:
--bugfix in StorageParser

CptHouston committed:
-- updated algorithm for relativeMonthly

CptHouston committed:
-- updated accessibility for non-public eventhandler methods

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandler.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace CharlieBackend.Business.Services.ScheduleServiceFolder
+{
+    public class ScheduledEventHandler : IScheduledEventHandler
```
Change this to Factory method class with the empty constructor ScheduleEventHandlerFactory.

CptHouston replied:
Updated

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandler.cs:
```diff
@@ -0,0 +1,46 @@
+    {
+        private IScheduledEventHandler _content;
+
+        public ScheduledEventHandler(EventOccurrence source, ContextForCreateScheduleDTO context, PatternForCreateScheduleDTO pattern)
```
Change this to method Create(PatternForCreateScheduleDTO )

CptHouston replied:
Updated, only pattern is passed to the constructor

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandler.cs:
```diff
@@ -0,0 +1,46 @@
+            switch (source.Pattern)
+            {
+                case PatternType.Daily:
+                    _content = new ScheduledEventHandlerDaily(source, context, pattern);
```
Use return instance here  to don't use break

CptHouston replied:
Updated

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandler.cs:
```diff
@@ -0,0 +1,46 @@
+            }
+        }
+
+        public IEnumerable<ScheduledEvent> GetEvents()
```
Use GetEvents from Handlers that will be returned by Get method

CptHouston replied:
Updated

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerRelativeMonthly.cs:
```diff
@@ -0,0 +1,72 @@
+{
+    public class ScheduledEventHandlerRelativeMonthly : ScheduledEventHandlerBase
+    {
+        public ScheduledEventHandlerRelativeMonthly(EventOccurrence source, ContextForCreateScheduleDTO context, PatternForCreateScheduleDTO pattern) 
```
Use only PatternForCreateScheduleDTO  as a parameter for this constructor. Other parameter please pass as an argument to the methods.

CptHouston replied:
Updated

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerBase.cs:
```diff
@@ -0,0 +1,76 @@
+            _pattern = pattern;
+        }
+
+        public virtual IEnumerable<ScheduledEvent> GetEvents()
```
make GetEvents(EventOccurrence , ContextForCreateScheduleDTO )

CptHouston replied:
updated

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/EventHandlers/ScheduledEventHandlerBase.cs:
```diff
@@ -0,0 +1,76 @@
+        public ScheduledEvent CreateEvent(DateTime targetStartDate, DateTime targetFinishDate,
+            EventOccurrence source, ContextForCreateScheduleDTO context)
+        {
+            return new ScheduledEvent
```
Inline this method.

CptHouston replied:
Updated

TetianaDomanska-dev reviewed (approved):
Looks good!

CptHouston committed:
-- updated scheduleEventHandler
(constructor now gets only pattern, rest is passed via params to GetEvents)
-- ScheduleEventFactory class created

mahotskuy commented on CharlieBackend.Business/Services/ScheduleServiceFolder/ScheduleService.cs:
```diff
@@ -0,0 +1,227 @@
+
+            _unitOfWork.EventOccurenceRepository.Add(result);
+
+            _unitOfWork.ScheduledEventRepository.AddRange(new ScheduledEventHandlerFactory().Get(createScheduleRequest.Pattern).GetEvents(result, createScheduleRequest.Context));
```
Please insert ISheduledEventHandlerFactory to the constructor dependency

CptHouston replied:
Updated

mahotskuy replied:
Please create an interface for ScheduledEventHandlerFactory and use DI to get implementation of ScheduledEventHandlerFactory. It will help us to mock data and simplify unit testing.

CptHouston committed:
-- IScheduledEventHandlerFactory dependency inverted

CptHouston committed:
Merge branch 'dev' into development/scheduleRework

CptHouston committed:
-- bugfix

CptHouston committed:
-- naming updated

CptHouston committed:
-- Added description for PostSchedule endpoint
